### PR TITLE
REFACTOR-#3900: add flake8-no-implicit-concat plugin and refactor flake8 error codes

### LINKED ISSUE #3900: Add linter to warn for python implicit concatenation
As seen in this [blog post](https://codereviewdoctor.medium.com/5-of-666-python-repos-had-comma-typos-including-tensorflow-and-pytorch-sentry-and-v8-7bc3ad9a1bb7), it's easy to make mistakes when using python implicit string concatenation (5% of repos, including popular ones like tensorflow, v8, pytorch make mistakes related to this). For example:

```
example_list = [
    "a",
    "b"
    "c"
]
# results in ["a", "bc"]
```

To avoid this we can try Code Review Doctor (paid) or try this flake8 rule: https://pypi.org/project/flake8-no-implicit-concat/.

### PR DESCRIPTION
**Pull request by jeffreykennethli**
## What do these changes do?
To avoid errors like this one: https://codereviewdoctor.medium.com/5-of-666-python-repos-had-comma-typos-including-tensorflow-and-pytorch-sentry-and-v8-7bc3ad9a1bb7, add a recommendation to use flake8-no-implicit-concat in the contributing docs and add this plugin to the CI.

- [ ] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [ ] Resolves #3900? 
- [ ] tests added and passing
- [ ] module layout described at `docs/developer/architecture.rst` is up-to-date

### CODE REVIEW

**Comment by mvashishtha** [comment]
Could you please try adding an implicit concat to this PR and checking whether the CI catches it?

**Review by devin-petersohn — COMMENTED** [review]
@jeffreykennethli I see some failures in CI formatting.

**Comment by dchigarev** [comment]
@jeffreykennethli
The `flake8` ignores warnings from the newly installed plugin because we explicitly list warning types that we're showing in Modin repo:
https://github.com/modin-project/modin/blob/fea20642ac30104fe2bc4cd7b616e756f2cfac23/setup.cfg#L24
We have to add a [`NIC` warning type](https://pypi.org/project/flake8-no-implicit-concat/) in the selection as well, to enable warnings from the plugin

**Comment by jeffreykennethli** [comment]
Thanks @dchigarev. I added the NIC error codes and the errors show up now.

I noticed we also have flake8-print installed but no error codes specified, so I'll add those too.

File: scripts/test/examples.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 # noqa: MD02
 """Function examples for docstring testing."""
 
+
```
[the remark this answers is not in the document]
Comment (jeffreykennethli) [reply scripts/test/examples.py:18]: This was formatted according to black's rules, see https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#empty-lines

[the remark this answers is not in the document]
Comment (devin-petersohn) [reply scripts/test/examples.py:18]: Let's make this change in a different PR.

**Review by mvashishtha — APPROVED** [review]

**Review by vnlitvinov — APPROVED** [review]
LGTM, but do note that concatenating strings via `+` has some performance penalties because this concatenation happens at runtime vs "implicit concatenation".

I think we can trade a bit of performance in exchange of better consistency here, though.

File: scripts/test/examples.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeffreykennethli) [reply scripts/test/examples.py]: Removed this change from this PR.

**Review by devin-petersohn — DISMISSED** [review]
LGTM

**Comment by devin-petersohn** [comment]
@modin-project/modin-omnisci This needs approval

**Review by Garra1980 — DISMISSED** [review]

**Review by YarShev — COMMENTED** [review]
Please add a release note. Otherwise, LGTM!

**Review by YarShev — APPROVED** [review]

**Pull request merged**